realm/realm-library/src/main/cpp/objectserver_shared.hpp

```diff
@@ -60,7 +61,19 @@ class JniSession {
                 coordinator->wake_up_notifier_worker();
             }
         };
-        auto error_handler = [weak_session_ref](std::error_code error_code, bool /*is_fatal*/, const std::string message) {
+        auto error_handler = [weak_session_ref](std::error_code error_code, bool is_fatal, const std::string message) {
+            if (error_code.category() != realm::sync::protocol_error_category() ||
+                    error_code.category() != realm::sync::client_error_category()) {
+                // FIXME: Consider below when moving to the OS sync manager.
```

Comment:
Wouldn't finding a way to kill the C++ thread after a Java exception fix this? It would mean we also need to find a way to restart the client (although I believe OS does that by itself), so sounds doable?